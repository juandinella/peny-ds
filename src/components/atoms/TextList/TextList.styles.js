import { css } from 'styled-components'

export default css`
  li {
    margin-bottom: var(--spacing-xxxs);

    &::marker {
      color: var(--color-primary-main);
    }
  }

  ${(props) =>
    props.inline &&
    css`
      li {
        display: inline-block;
        margin-right: var(--spacing-xs);
      }
    `}

  ${(props) =>
    props.color &&
    css`
      li {
        color: var(--color-${(props) => props.color});
      }
    `}

  ${(props) =>
    props.bulletColor &&
    css`
      li {
        &::marker {
          color: var(--color-${(props) => props.bulletColor});
        }
      }
    `}


  ${(props) =>
    props.size &&
    css`
      li {
        font-size: var(--font-${(props) => props.size});
      }
    `}
`
