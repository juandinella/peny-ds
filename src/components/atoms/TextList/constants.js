export const options = {
  /* Size of the list */
  sizes: ['xxs', 'xs', 'sm', 'md', 'lg', 'xl', 'xxl', 'xxxl'],
  /* Color of the list */
  colors: [
    'neutral-900',
    'neutral-800',
    'neutral-700',
    'neutral-600',
    'neutral-500',
    'neutral-400',
    'neutral-300',
    'neutral-200',
    'neutral-100',
    'neutral-50',
    'neutral-00',
    'primary-main',
    'secondary-main',
  ],
}
