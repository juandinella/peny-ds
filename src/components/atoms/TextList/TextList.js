import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import styles from './TextList.styles'
import Skeleton from '../Skeleton'
import Utils from '../../../utils'
import { options } from './constants'

/**
 * List of items
 */

const TextList = ({
  className,
  itemClass,
  list,
  id,
  inline,
  color,
  bulletColor,
  numbered,
  size,
  shouldRender,
}) => {
  const texts = list
    ? list.map((item, index) => (
        <Fragment key={index}>
          {item.content ? (
            <li
              className={`${itemClass || ''} ${item.className || ''}`}
              id={item.id}
              data-testid={item.id}
              onClick={item.onClick || null}
            >
              {item.content}
            </li>
          ) : null}
          {item.children}
        </Fragment>
      ))
    : []

  return (
    <Skeleton
      id={`${id}_skeleton`}
      show={!Utils.emptyValue(shouldRender) && !shouldRender()}
    >
      {numbered ? (
        <ol
          id={id}
          css={styles}
          data-testid={id}
          className={className}
          inline={inline}
          size={size}
          color={color}
          bulletColor={bulletColor}
        >
          {texts}
        </ol>
      ) : (
        <ul
          id={id}
          css={styles}
          data-testid={id}
          className={className}
          inline={inline}
          size={size}
          color={color}
          bulletColor={bulletColor}
        >
          {texts}
        </ul>
      )}
    </Skeleton>
  )
}

TextList.displayName = 'TextList'

TextList.propTypes = {
  /** To add style */
  className: PropTypes.string,
  /** TextList's id */
  id: PropTypes.string,
  /** Item list class */
  itemClass: PropTypes.string,
  /** Numbered list */
  numbered: PropTypes.bool,
  /** Array of objects for list */
  list: PropTypes.arrayOf(
    PropTypes.shape({
      /** Content for the list item. */
      content: PropTypes.oneOfType([PropTypes.string, PropTypes.node])
        .isRequired,
      /** Specifies a Unique id to the item list element. */
      id: PropTypes.string,
      /** Extends the styles applied to the list item.*/
      className: PropTypes.string,
      /** Callback fired when item list is clicked. */
      onClick: PropTypes.func,
      /** Nodes after list item. */
      children: PropTypes.node,
    })
  ),
  /**
   * Function that decides if the component should be rendered.
   *
   * @return {boolean} false if component should not be rendered and show skeleton instead. Component renders otherwise.
   */
  shouldRender: PropTypes.func,
  /** Inline list */
  inline: PropTypes.bool,
  /** Size of the text */
  size: PropTypes.oneOf(options.sizes),
  /** Color of the text */
  color: PropTypes.oneOf(options.colors),
  /** Color of the bullet */
  bulletColor: PropTypes.oneOf(options.colors),
}

TextList.defaultProps = {}

export default TextList
