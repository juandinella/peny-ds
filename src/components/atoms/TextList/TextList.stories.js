import TextList from './TextList'
import { options } from './constants'
import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(TextList)
const ListTemplate = getListTemplate(TextList)

export default {
  title: 'Atoms/TextList',
  component: TextList,
  args: {
    inline: false,
    list: [{ content: 'First Item' }, { content: 'Second Item' }],
  },
  argTypes: {
    color: getOptionsArgTypes(options.colors),
    bulletColor: getOptionsArgTypes(options.colors),
    size: getOptionsArgTypes(options.sizes),
  },
}

export const Default = Template.bind({})

export const Skeleton = Template.bind({})
Skeleton.args = { shouldRender: () => false }

export const Colors = ListTemplate.bind({})
Colors.args = { items: options.colors.map((color) => ({ color })) }

export const Numbered = Template.bind({})
Numbered.args = { numbered: true, bulletColor: 'neutral-900' }

export const BulletColors = ListTemplate.bind({})
BulletColors.args = {
  items: options.colors.map((bulletColor) => ({ bulletColor })),
}

export const Sizes = ListTemplate.bind({})
Sizes.args = { items: options.sizes.map((size) => ({ size })) }
