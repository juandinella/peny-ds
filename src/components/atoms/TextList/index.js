import TextList from './TextList'

export default TextList
export { options } from './constants'
