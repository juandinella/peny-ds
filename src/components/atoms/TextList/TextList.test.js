import React from 'react'

import { cleanup, render, fireEvent } from '@testing-library/react'

import TextList from './TextList'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Atoms\/TextList$/,
})

afterEach(cleanup)

const componentProps = {
  list: [
    {
      content: 'First Item',
      onClick: jest.fn(),
      className: 'custom_first',
    },
    {
      content: 'Second Item',
      onClick: jest.fn(),
      className: 'custom_second',
    },
  ],
}

beforeEach(() => {
  componentProps.list[0].onClick.mockClear()
  componentProps.list[1].onClick.mockClear()
})

describe('<TextList />', () => {
  test('should render texts correctly', () => {
    const { container } = render(<TextList {...componentProps} />)

    const listChildrens = container.querySelectorAll('ul > li')

    expect(listChildrens).toHaveLength(2)

    expect(listChildrens[0].textContent).toEqual('First Item')
    expect(listChildrens[1].textContent).toEqual('Second Item')
  })

  test('should render skeleton', () => {
    const { queryByTestId } = render(
      <TextList {...componentProps} shouldRender={() => false} />
    )

    expect(queryByTestId(`${componentProps.id}_skeleton`)).toBeTruthy()
  })

  test('should render class correctly', () => {
    const { container, rerender } = render(
      <TextList {...componentProps} className="custom_class" />
    )

    const list = container.firstChild

    expect(list.className).toContain('custom_class')

    rerender(<TextList {...componentProps} className={null} />)

    expect(list.className).not.toContain('custom_class')
  })

  test('should render item children correctly', () => {
    const { container, rerender } = render(<TextList {...componentProps} />)

    expect(container.querySelector('ul').children.length).toEqual(2)

    rerender(
      <TextList
        {...componentProps}
        list={[
          { content: 'First Item', children: <div>Children</div> },
          { content: 'Second Item' },
        ]}
      />
    )

    const items = container.querySelector('ul').children

    expect(items.length).toEqual(3)
    expect(items[1].textContent).toEqual('Children')
  })

  test('should render item class correctly', () => {
    const { container, rerender } = render(
      <TextList {...componentProps} itemClass="custom_class" />
    )

    const items = container.querySelectorAll('li')

    expect(items[0].className).toContain('custom_class')
    expect(items[1].className).toContain('custom_class')

    rerender(<TextList {...componentProps} itemClass={null} />)

    expect(items[0].className).not.toContain('custom_class')
    expect(items[1].className).not.toContain('custom_class')
  })

  test('item should be clicked', () => {
    const { container, rerender } = render(<TextList {...componentProps} />)

    const items = container.querySelectorAll('li')

    fireEvent.click(items[0])
    fireEvent.click(items[1])

    expect(componentProps.list[0].onClick).toHaveBeenCalledTimes(1)
    expect(componentProps.list[1].onClick).toHaveBeenCalledTimes(1)

    rerender(
      <TextList
        {...componentProps}
        list={[{ content: 'First Item' }, { content: 'Second Item' }]}
      />
    )

    fireEvent.click(items[0])
    fireEvent.click(items[1])

    expect(componentProps.list[0].onClick).toHaveBeenCalledTimes(1)
    expect(componentProps.list[1].onClick).toHaveBeenCalledTimes(1)
  })
})
