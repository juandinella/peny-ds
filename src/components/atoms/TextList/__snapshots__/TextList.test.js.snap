// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/TextList Bullet Colors 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <ul
    className="sc-2bi9jj-1 gMGmnE"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 eEYlnV"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 kDdDri"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 ivvCrz"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 kpUdZc"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 iimcZt"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 hohOve"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 fgzNvv"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 haYpcY"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 DjsTA"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 gwbmwJ"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 dUptlD"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 bXbLUx"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
</div>
`;

exports[`Storyshots Atoms/TextList Colors 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <ul
    className="sc-2bi9jj-1 yBhfm"
    color="neutral-900"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 QaOmX"
    color="neutral-800"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 ioRYkM"
    color="neutral-700"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 efdVgh"
    color="neutral-600"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 kZEhHO"
    color="neutral-500"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 bIfqD"
    color="neutral-400"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 cFwztE"
    color="neutral-300"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 jLefNV"
    color="neutral-200"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 fqiIQG"
    color="neutral-100"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 ceLigK"
    color="neutral-50"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 cRFoqz"
    color="neutral-00"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 bKDXXN"
    color="primary-main"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 ljPgv"
    color="secondary-main"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
</div>
`;

exports[`Storyshots Atoms/TextList Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <ul
    className="sc-2bi9jj-1 gEomZU"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
</div>
`;

exports[`Storyshots Atoms/TextList Numbered 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <ol
    className="sc-2bi9jj-0 fWUvtx"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ol>
</div>
`;

exports[`Storyshots Atoms/TextList Sizes 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <ul
    className="sc-2bi9jj-1 iozkJE"
    size="xxs"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 fUymAA"
    size="xs"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 jWqudB"
    size="sm"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 esyglK"
    size="md"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 dkWtIk"
    size="lg"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 idoWof"
    size="xl"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 fXHkfH"
    size="xxl"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
  <ul
    className="sc-2bi9jj-1 chxIFX"
    size="xxxl"
  >
    <li
      className=" "
      onClick={null}
    >
      First Item
    </li>
    <li
      className=" "
      onClick={null}
    >
      Second Item
    </li>
  </ul>
</div>
`;

exports[`Storyshots Atoms/TextList Skeleton 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-kjnynt-0 jfZhXD"
    data-testid="undefined_skeleton"
    id="undefined_skeleton"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <ul
      className="sc-2bi9jj-1 gEomZU"
    >
      <li
        className=" "
        onClick={null}
      >
        First Item
      </li>
      <li
        className=" "
        onClick={null}
      >
        Second Item
      </li>
    </ul>
  </div>
</div>
`;
